src/main/drivers/dshot_dpwm.c

```diff
@@ -0,0 +1,196 @@
+ * will be useful, but WITHOUT ANY WARRANTY; without even the implied
+ * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this software.
+ *
+ * If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Author: jflyper
+ */
+
+#include <stdbool.h>
+#include <stdint.h>
+#include <math.h>
+#include <string.h>
+
+#include "platform.h"
+
+#ifdef USE_DSHOT
+
+#include "drivers/pwm_output.h"
+#include "drivers/dshot.h"
+#include "drivers/dshot_dpwm.h"
+#include "drivers/motor.h"
+
+#include "pg/motor.h"
+
+// XXX TODO: Share a single region among dshotDmaBuffer and dshotBurstDmaBuffer
+
+DSHOT_DMA_BUFFER_ATTRIBUTE DSHOT_DMA_BUFFER_UNIT dshotDmaBuffer[MAX_SUPPORTED_MOTORS][DSHOT_DMA_BUFFER_ALLOC_SIZE];
+
+#ifdef USE_DSHOT_DMAR
+DSHOT_DMA_BUFFER_ATTRIBUTE DSHOT_DMA_BUFFER_UNIT dshotBurstDmaBuffer[MAX_DMA_TIMERS][DSHOT_DMA_BUFFER_SIZE * 4];
+#endif
+
+#ifdef USE_DSHOT_DMAR
+FAST_RAM_ZERO_INIT bool useBurstDshot = false;
+#endif
+#ifdef USE_DSHOT_TELEMETRY
+FAST_RAM_ZERO_INIT bool useDshotTelemetry = false;
+#endif
+
+FAST_RAM_ZERO_INIT loadDmaBufferFn *loadDmaBuffer;
+
+FAST_CODE uint8_t loadDmaBufferDshot(uint32_t *dmaBuffer, int stride, uint16_t packet)
+{
+    for (int i = 0; i < 16; i++) {
+        dmaBuffer[i * stride] = (packet & 0x8000) ? MOTOR_BIT_1 : MOTOR_BIT_0;  // MSB first
+        packet <<= 1;
+    }
+    dmaBuffer[16 * stride] = 0;
+    dmaBuffer[17 * stride] = 0;
+
+    return DSHOT_DMA_BUFFER_SIZE;
+}
+
+FAST_CODE uint8_t loadDmaBufferProshot(uint32_t *dmaBuffer, int stride, uint16_t packet)
+{
+    for (int i = 0; i < 4; i++) {
+        dmaBuffer[i * stride] = PROSHOT_BASE_SYMBOL + ((packet & 0xF000) >> 12) * PROSHOT_BIT_WIDTH;  // Most significant nibble first
+        packet <<= 4;   // Shift 4 bits
+    }
+    dmaBuffer[4 * stride] = 0;
+    dmaBuffer[5 * stride] = 0;
+
+    return PROSHOT_DMA_BUFFER_SIZE;
+}
+
+uint32_t getDshotHz(motorPwmProtocolTypes_e pwmProtocolType)
+{
+    switch (pwmProtocolType) {
+    case(PWM_TYPE_PROSHOT1000):
+        return MOTOR_PROSHOT1000_HZ;
+    case(PWM_TYPE_DSHOT1200):
+        return MOTOR_DSHOT1200_HZ;
+    case(PWM_TYPE_DSHOT600):
+        return MOTOR_DSHOT600_HZ;
+    case(PWM_TYPE_DSHOT300):
+        return MOTOR_DSHOT300_HZ;
+    default:
+    case(PWM_TYPE_DSHOT150):
+        return MOTOR_DSHOT150_HZ;
+    }
+}
+
+static void dshotPwmShutdown(void)
+{
+    // DShot signal is only generated if write to motors happen,
+    // and that is prevented by enabled checking at write.
+    // Hence there's no special processing required here.
+    return;
+}
+
+static void dshotPwmDisableMotors(void)
+{
+    // No special processing required
+    return;
+}
+
+static bool dshotPwmEnableMotors(void)
+{
+    // No special processing required
+    return true;
+}
+
+extern void pwmWriteDshotInt(uint8_t index, uint16_t value); // XXX Do this right
```

Comment:
This must go into some header file. Changing of function signature will not be caught by compiler and lead to very nasty bugs

Reply:
Do this right!!! Yes!